core/src/tile/tileManager.cpp

```diff
@@ -303,204 +215,244 @@ void TileManager::updateTileSet(TileSet& tileSet) {
                 //     NOT_A_TILE. (for the current implementation of > operator)
                 assert(!(visTileId == NOT_A_TILE));
 
-                // tileSet is missing an element present in visibleTiles
-                if (!addTile(tileSet, visTileId)) {
+                if (!addTile(_tileSet, visTileId)) {
                     // Not in cache - enqueue for loading
-                    enqueueTask(tileSet, visTileId, viewCenter);
+                    enqueueTask(_tileSet, visTileId, viewCenter);
                 }
 
                 ++visTilesIt;
 
             } else {
-                assert(!(curTileId == NOT_A_TILE));
                 // tileSet has a tile not present in visibleTiles
-                auto& tile = curTilesIt->second;
+                assert(!(curTileId == NOT_A_TILE));
+                auto& entry = curTilesIt->second;
 
-                if (!tile) {
-                    curTilesIt = tiles.erase(curTilesIt);
-                    continue;
-                }
+                if (entry.getProxyCounter() > 0) {
+                    if (entry.newData()) {
+                        clearProxyTiles(_tileSet, curTileId, entry, removeTiles);
+                        entry.tile = std::move(entry.task->tile());
+                        entry.task.reset();
+                    }
+                    if (entry.isReady()) {
+                        m_tiles.push_back(entry.tile);
 
-                if (tile->getProxyCounter() <= 0) {
-                    removeTiles.push_back(tile->getID());
-                } else if (tile->isReady()) {
-                    m_tiles.push_back(tile);
-                } else if (tile->getID().z > maxZoom) {
-                    // cancel requsts for tiles that
-                    // are too small on screen
-                    removeTiles.push_back(tile->getID());
+                    } else if (curTileId.z < maxZoom){
+                        removeTiles.push_back(curTileId);
+                    }
+                } else {
+                    removeTiles.push_back(curTileId);
                 }
-                tile->setVisible(false);
-
                 ++curTilesIt;
             }
         }
     }
 
     while (!removeTiles.empty()) {
-        auto tileIt = tiles.find(removeTiles.back());
+        auto it = tiles.find(removeTiles.back());
         removeTiles.pop_back();
 
-        if ((tileIt != tiles.end()) &&
-            (tileIt->second->getProxyCounter() <= 0 ||
-             tileIt->second->getID().z > maxZoom)) {
+        if ((it != tiles.end()) &&
+            (it->second.getProxyCounter() <= 0  ||
+             it->first.z >= maxZoom)) {
+
+            auto& entry = it->second;
+
+            LOG("REMOVE %s - ready:%d proxy:%d/%d",
+                it->first.toString().c_str(),
+                entry.isReady(),
+                entry.getProxyCounter(),
+                entry.m_proxies);
+
+            if (visibleTiles.find(it->first) != visibleTiles.end()) {
```

Comment:
Yes, there may be a better way to solve this. Currently it is possible to have circles in the proxies where a parent references its children and the children reference their parent as proxies. This can lead to adding current tiles to the removal list.